### DISCUSSION TITLE
Apply default value to widgets

### DISCUSSION THREAD

**Discussion by sdrenth** [algolia/instantsearch#4758] [Q&A]
What's described here is basically what I try to achieve:
https://www.algolia.com/doc/guides/building-search-ui/widgets/customize-an-existing-widget/js/#apply-default-value-to-widgets

I want to have a default, preselected value of a menu widget (instantsearch.widgets.menu). I have two types of results, products and pages. By default I want to only list results of type products. Only when the visitor selects the Pages type I want to list the pages.

I do this by setting: `disjunctiveFacetsRefinements.facet = ['Products'];`

Then only products are listed on initial page load (as expected), however I run into two issues:

1. The filter option is not shown as preselected (only after clicking the filter option):
![Screenshot 2021-05-20 at 17 50 43](https://user-images.githubusercontent.com/7802465/119010056-f82d9500-b993-11eb-8501-0e41e066f391.png)
2. When I click the Pages option, no results are shown and the filter options disappear

**Comment by Haroenv** [comment]
You can set `initialUiState` on the InstantSearch instance: https://codesandbox.io/s/thirsty-curie-vibvi?file=/src/app.js, is that what you're looking for? Or is it defined from the URL (in which case a router works best)

**Reply by sdrenth** [reply]
@Haroenv 

> Or is it defined from the URL (in which case a router works best)

 No, its not based on URL parameters.

I also tried with the initialUiState and it works fine with refinementList but I need it to work with menu, because it behaves similarly to radio buttons:
```
initialUiState: {
    dev_main: {
        menu: {
            facet: ['Products'],
        },
    },
},

instantsearch.widgets.menu({
    container       : '.is-filters-facet',
    attribute       : 'facet',
    limit           : 5,
    showMore        : false,
    searchable      : false,
    sortBy          : ['name:asc'],
    templates       : {
        item                : this.renderFilterItem
    },
    cssClasses      : this.refinementCSSClasses
})
```

However this returns an error:
```
scripts.js:41549 Uncaught TypeError: n.split is not a function
    at scripts.js:41549
    at Array.reduce (<anonymous>)
    at Object._getHitsHierarchicalFacetsAttributes (scripts.js:41549)
    at Object._getHitsSearchParams (scripts.js:41549)
    at Object._getQueries (scripts.js:41549)
    at scripts.js:41549
    at Array.map (<anonymous>)
    at ee._search (scripts.js:41549)
    at ee.searchOnlyWithDerivedHelpers (scripts.js:41549)
    at ee.t.search (scripts.js:41549)
```

Also I would need the filter to always have an active filter item, just like the numericalMenu widget has in this example:
https://www.algolia.com/doc/guides/building-search-ui/widgets/showcase/js/

**Reply by Haroenv** [reply]
for menu it's not in an array :) `menu: { facet: 'Products' }`, since they can only have one value

**Reply by sdrenth** [reply]
@Haroenv ah ofcourse, thanks!

**Reply by sdrenth** [reply]
@Haroenv It is preselected now by default which is good, but I noticed a few things.

1. I'm able to deselect a filter item (unlike regular radio button behaviour). 1 value should be selected all the time. I tried adding a regular on change listener, but then the checked state has already changed. How can I prevent a user from unchecking an active filter item?
2.  When I search on a value which has no results, no filter items are shown which is default behaviour in instantsearch. I want users to always be able to switch between the facets even if there are no matches. I found this in the documentation, which works in terms of displaying facets with no results. However, then I'm able to select multiple values (so it behaves as a regular refinementlist) and gives really awkward behaviour. (As you can see in the video)
3. Ideally, I'd like to show a total count of results independent of an active facet. For example, when I search for Checkout and have Product as an active facet, I'd like to see a filter count for facet Pages indicating that when selecting that filter option the visitor will see a result. 

https://user-images.githubusercontent.com/7802465/119218913-e0185b80-bae2-11eb-90ab-03dbdc8486bd.mov

As you can see in the end of the video, the Pages filter option is active (as expected), but returns no results (where there should be one result) and the URL filters on Products facet:
?dev_main%5Bmenu%5D%5Bfacet%5D=Products&dev_main%5Bquery%5D=Checkout

**Reply by francoischalifour** [reply]
1. It sounds like what you want is not an initial refinement but a default refinement. If the value should't be always selected, I wouldn't display it at all, and reflect that in the title of the page. If you still want it to show, you can add the `disabled` HTML attribute.

2. It seems like the widget you want is a `menu`, and not a `refinementList`. Again, I would challenge the UX decision and either disable or hide (which is the default behavior) the category.

3. 

> I'd like to see a filter count for facet Pages indicating that when selecting that filter option the visitor will see a result

This is supposed to be the default behavior.

---

So we make sure to not talk past each other, it'd be more productive to reproduce what you currently have in [this sandbox](https://codesandbox.io/s/github/algolia/create-instantsearch-app/tree/templates/instantsearch.js) and try to get to the behavior you're looking for there.

**Reply by sdrenth** [reply]
@francoischalifour How do I set the default refinement for a menu?

[sandbox](https://codesandbox.io/s/instantsearch-8syr3?file=/src/index.js)

**Reply by sdrenth** [reply]
@Haroenv @francoischalifour Any updates on this one?

**Reply by francoischalifour** [reply]
You can use a combination of `initialUiState` and a virtual widget (i.e., a widget that doesn't render anything) to force a default refinement:

```js
const indexName = 'instant_search';
const defaultCategory = 'Smart Home';

const search = instantsearch({
  indexName,
  searchClient,
  initialUiState: {
    [indexName]: {
      menu: {
        categories: defaultCategory,
      },
    },
  },
});

const virtualMenu = instantsearch.connectors.connectMenu(() => {});

search.addWidgets([
  virtualMenu({
    attribute: 'categories',
  }),
]);
```

Try it in [that sandbox](https://codesandbox.io/s/instantsearchjs-default-category-6743f?file=/src/app.js).

**Reply by sdrenth** [reply]
@francoischalifour I tried (by editing the sandbox to my needs), but this still gives unexpected behaviour. 

Please see my [sandbox](https://codesandbox.io/s/instantsearchjs-default-category-forked-99q7n?file=/src/app.js):

On initial load it appears to be okay, but as soon as I select a different filter from the menu "Smart Home" disappears from the menu. :pensive:

**Reply by francoischalifour** [reply]
(1) If you want to always fall back to a default category, you may need to use [`connectMenu`](https://www.algolia.com/doc/api-reference/widgets/menu/js/#connector).

(2) If you still want to allow other refinements on top of that default refinement, you may need to use [`connectRefinementList`](https://www.algolia.com/doc/api-reference/widgets/refinement-list/js/#connector).

**Reply by Haroenv** [reply]
something interesting can also be the [onStateChange](https://www.algolia.com/doc/api-reference/widgets/instantsearch/js/#widget-param-onstatechange) method, this will allow you to keep a certain refinement enabled, even if someone wants to disable it for example, it's not really default or initial refinement, but rather forced refinement you want I think

**Reply by sdrenth** [reply]
I think it's odd that this is such a hassle for something this basic. But basically, I need to write my own logic using connectMenu, I'll try that or I'll start looking for a different work-around, because this has taken way to much time now.

**Reply by Haroenv** [reply]
Do you have any examples of other websites with this behaviour? I've never seen filters that you can't remove

**Reply by sdrenth** [reply]
@Haroenv The only thing I need is basically a HTML radio element with a predefined value. Once a radio element has a value selected you cannot unselect it as well.

**Reply by Haroenv** [reply]
I still really don't understand what you want, is it this? https://codesandbox.io/s/thirsty-curie-vibvi?file=/src/app.js

**Reply by sdrenth** [reply]
@Haroenv Basically yes, I can make it a radio and just style it differently. However, using your codesandbox I cannot switch to a different brand.

**Reply by Haroenv** [reply]
I'm completely lost, can you give an example website that behaves how you want please? I don't understand which options aren't yet given in this thread

**Reply by sdrenth** [reply]
@Haroenv I can't give you an example website that behaves like how I like it to, because I wouldn't know a website which would have behaviour like this. It comes down to this (on pageload):
```
<input type="radio" id="male" name="gender" value="male" checked>
<label for="male">Male</label><br>
<input type="radio" id="female" name="gender" value="female">
<label for="female">Female</label><br>
<input type="radio" id="other" name="gender" value="other">
<label for="other">Other</label>
```

So what I like to achieve is exactly like this sandbox here if it would be working properly:
[https://codesandbox.io/s/thirsty-curie-vibvi?file=/src/app.js](https://codesandbox.io/s/thirsty-curie-vibvi?file=/src/app.js)

So, just like your example, we have a predefined value on pageload ("Samsung") and the visitor should be able to switch between brands (currently not working in your sandbox) but should not be able to undo the filter.

https://user-images.githubusercontent.com/7802465/121212743-e8f88380-c87d-11eb-89b0-5738c2e40f6a.mov

Above is exactly what I'd like to achieve, I tried doing something like this with out-of-the-box solutions such as the menu refinement, but that doesn't seem to work.

**Reply by Haroenv** [reply]
so is this what you want? https://codesandbox.io/s/thirsty-curie-vibvi?file=/src/app.js (edited)

**Reply by sdrenth** [reply]
@Haroenv Yes thanks, that will do as well! I'll just style it so it looks like a regular menu ;)
